File: src/test/java/tech/sirwellington/alchemy/http/HttpAssertionsTest.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http;

import java.net.MalformedURLException;
import java.net.URI;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import tech.sirwellington.alchemy.arguments.FailedAssertionException;
import tech.sirwellington.alchemy.test.AlchemyTest;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.notNullValue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.generator.NumberGenerators.integers;
import static tech.sirwellington.alchemy.generator.StringGenerators.*;
import static tech.sirwellington.alchemy.test.ThrowableAssertion.assertThrows;

/**
 * @author SirWellington
 */
@AlchemyTest
public class HttpAssertionsTest {

    private RequestMethod requestMethod;

    @BeforeEach
    public void setUp() {
    }

    @Test
    public void testConstructor() {
        assertThrows(
            () -> HttpAssertions.class.getDeclaredConstructor().newInstance()
        );
    }

    @Test
    public void testValidHttpStatusCode() {
        var instance = HttpAssertions.validHttpStatusCode();
        assertThat(instance, notNullValue());

        var statusCode = one(integers(200, 500));
        instance.check(statusCode);

        var badStatusCode = one(integers(-100, 100));
        assertThrows(() -> instance.check(badStatusCode))
            .isInstanceOf(FailedAssertionException.class);

        var anotherBadCode = one(integers(600, Integer.MAX_VALUE));
        assertThrows(() -> instance.check(anotherBadCode))
            .isInstanceOf(FailedAssertionException.class);
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testValidResponseClass() {
        // Check Object
        var instanceOne = HttpAssertions.validResponseClass();
        assertThat(instanceOne, notNullValue());
        instanceOne.check(Object.class);

        // Check String
        var instanceTwo = HttpAssertions.validResponseClass();
        instanceTwo.check((Class<Object>) (Class<?>) String.class);

        // Edge Cases
        assertThrows(() -> instanceOne.check(null))
            .isInstanceOf(FailedAssertionException.class);

        var instanceThree = HttpAssertions.validResponseClass();
        assertThrows(() -> instanceThree.check((Class<Object>) (Class<?>) Void.class))
            .isInstanceOf(FailedAssertionException.class);
    }

    @Test
    public void testRequestReady() throws MalformedURLException {
        var instance = HttpAssertions.ready();

        var url = one(Generators.validUrls());

        var request = mock(HttpRequest.class);
        when(request.url()).thenReturn(url);
        when(request.method()).thenReturn(requestMethod);

        instance.check(request);
    }

    @Test
    public void testRequestReadyEdgeCases() throws Exception {
        var instance = HttpAssertions.ready();

        // Edge cases
        assertThrows(() -> instance.check(null))
            .isInstanceOf(FailedAssertionException.class);

        var url = one(Generators.validUrls());
        var request = mock(HttpRequest.class);
        when(request.url()).thenReturn(url);
        when(request.method()).thenReturn(requestMethod);

        // Missing Request Method
        when(request.method()).thenReturn(null);

        assertThrows(() -> instance.check(request))
            .isInstanceOf(FailedAssertionException.class);

        when(request.method()).thenReturn(requestMethod);

        // Missing URL
        when(request.url()).thenReturn(null);

        assertThrows(() -> instance.check(request))
            .isInstanceOf(FailedAssertionException.class);

        // Bad URL
        var badUrl = new URI("file://" + one(alphabeticStrings()));
        when(request.url()).thenReturn(badUrl.toURL());

        assertThrows(() -> instance.check(request))
            .isInstanceOf(FailedAssertionException.class);
    }

    @Test
    public void testValidContentType() {
        var instance = HttpAssertions.validContentType();
        assertThat(instance, notNullValue());

        var contentType = one(stringsFromFixedList(
            ContentTypes.APPLICATION_JSON,
            ContentTypes.PLAIN_TEXT
        ));

        instance.check(contentType);
        instance.check(contentType + one(alphabeticStrings()));
    }

    @Test
    public void testValidContentTypeEdgeCases() {
        var instance = HttpAssertions.validContentType();

        // Edge cases
        assertThrows(() -> instance.check(null))
            .isInstanceOf(FailedAssertionException.class);

        assertThrows(() -> instance.check(""))
            .isInstanceOf(FailedAssertionException.class);

        assertThrows(() -> instance.check(one(alphabeticStrings())))
            .isInstanceOf(FailedAssertionException.class);

        assertThrows(() -> instance.check(one(hexadecimalString(10))))
            .isInstanceOf(FailedAssertionException.class);
    }

    @Test
    public void testJsonArray() {
        var instance = HttpAssertions.jsonArray();
        assertThat(instance, notNullValue());

        var valid = one(Generators.jsonArrays());
        instance.check(valid);

        var object = one(Generators.jsonObjects());
        assertThrows(() -> instance.check(object))
            .isInstanceOf(FailedAssertionException.class);

        var primitive = one(Generators.jsonPrimitives());
        assertThrows(() -> instance.check(primitive))
            .isInstanceOf(FailedAssertionException.class);
    }

    @Test
    public void testOkResponse() {
        var instance = HttpAssertions.okResponse();
        assertThat(instance, notNullValue());

        // Check with null argument
        assertThrows(() -> instance.check(null))
            .isInstanceOf(FailedAssertionException.class);

        // Response is OK
        var okResponse = mock(HttpResponse.class);
        when(okResponse.isOk()).thenReturn(true);
        instance.check(okResponse);

        // Response is NOT OK
        var notOkResponse = mock(HttpResponse.class);
        when(notOkResponse.isOk()).thenReturn(false);
        assertThrows(() -> instance.check(notOkResponse))
            .isInstanceOf(FailedAssertionException.class);
    }
}

File: src/test/java/tech/sirwellington/alchemy/http/HttpAssertionsTest.kt
/*
 * Copyright © 2019. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http

import com.nhaarman.mockito_kotlin.whenever
import org.hamcrest.Matchers.notNullValue
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import tech.sirwellington.alchemy.arguments.FailedAssertionException
import tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.integers
import tech.sirwellington.alchemy.generator.StringGenerators
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.alphabeticStrings
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.hexadecimalString
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.AlchemyTestRunner
import tech.sirwellington.alchemy.test.junit.runners.GenerateEnum
import tech.sirwellington.alchemy.test.junit.runners.Repeat
import java.net.MalformedURLException
import java.net.URL

/**
 *
 * @author SirWellington
 */
@RunWith(AlchemyTestRunner::class)
class HttpAssertionsTest
{

    @GenerateEnum
    private val requestMethod: RequestMethod? = null

    @Before
    fun setUp()
    {
    }

    @Test
    fun testConstructor()
    {
        assertThrows { HttpAssertions::class.java.newInstance() }
                .isInstanceOf(IllegalAccessException::class.java)
    }

    @Repeat
    @Test
    fun testValidHttpStatusCode()
    {
        val instance = HttpAssertions.validHttpStatusCode()
        assertThat(instance, notNullValue())

        val statusCode = one(integers(200, 500))
        instance.check(statusCode)

        val badStatusCode = one(integers(-100, 100))

        assertThrows { instance.check(badStatusCode) }
                .isInstanceOf(FailedAssertionException::class.java)

        val anotherBadCode = one(integers(600, Int.MAX_VALUE))
        assertThrows { instance.check(anotherBadCode) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testValidResponseClass()
    {
        //Check Object
        val instanceOne = HttpAssertions.validResponseClass<Any>()
        assertThat(instanceOne, notNullValue())
        instanceOne.check(Any::class.java)

        //Check String
        val instanceTwo = HttpAssertions.validResponseClass<String>()
        instanceTwo.check(String::class.java)

        //Edge Cases
        assertThrows { instanceOne.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        val instanceThree = HttpAssertions.validResponseClass<Void>()
        assertThrows { instanceThree.check(Void::class.java) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    @Throws(MalformedURLException::class)
    fun testRequestReady()
    {
        val instance = HttpAssertions.ready()

        val url = one(Generators.validUrls())

        val request = mock(HttpRequest::class.java)
        whenever(request.url).thenReturn(url)
        whenever(request.method).thenReturn(requestMethod)

        instance.check(request)
    }

    @Repeat
    @Test
    @Throws(MalformedURLException::class)
    fun testRequestReadyEdgeCases()
    {
        val instance = HttpAssertions.ready()

        //Edge cases
        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        val url = one(Generators.validUrls())
        val request = mock(HttpRequest::class.java)
        whenever(request.url).thenReturn(url)
        whenever(request.method).thenReturn(requestMethod)

        //Missing Request Method
        whenever(request.method).thenReturn(null)

        assertThrows { instance.check(request) }
                .isInstanceOf(FailedAssertionException::class.java)

        whenever(request.method).thenReturn(requestMethod)

        //Missing URL
        whenever(request.url).thenReturn(null)

        assertThrows { instance.check(request) }
                .isInstanceOf(FailedAssertionException::class.java)

        //Bad URL
        val badUrl = URL("file://" + one(alphabeticStrings()))
        whenever(request.url).thenReturn(badUrl)

        assertThrows { instance.check(request) }
                .isInstanceOf(FailedAssertionException::class.java)

    }

    @Repeat(10)
    @Test
    fun testValidContentType()
    {
        val instance = HttpAssertions.validContentType()
        assertThat(instance, notNullValue())

        val validTypes = StringGenerators.stringsFromFixedList(ContentTypes.APPLICATION_JSON, ContentTypes.PLAIN_TEXT)

        val contentType = one(validTypes)

        instance.check(contentType)
        instance.check(contentType + one<String>(alphabeticStrings()))
    }

    @Test
    fun testValidContentTypeEdgeCases()
    {
        val instance = HttpAssertions.validContentType()

        //Edge cases
        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check("") }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(one(alphabeticStrings())) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(one(hexadecimalString(10))) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Repeat
    @Test
    fun testJsonArray()
    {
        val instance = HttpAssertions.jsonArray()
        assertThat(instance, notNullValue())

        val valid = one(Generators.jsonArrays())
        instance.check(valid)

        val `object` = one(Generators.jsonObjects())
        assertThrows { instance.check(`object`) }
                .isInstanceOf(FailedAssertionException::class.java)

        val primitive = one(Generators.jsonPrimitives())
        assertThrows { instance.check(primitive) }
                .isInstanceOf(FailedAssertionException::class.java)

    }

    @Test
    fun testOkResponse()
    {
        val instance = HttpAssertions.okResponse()
        assertThat(instance, notNullValue())

        //Check with null argument
        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        //Response is OK
        val okResponse = mock(HttpResponse::class.java)
        whenever(okResponse.isOk).thenReturn(true)
        instance.check(okResponse)

        //Response is NOT OK
        val notOkResponse = mock(HttpResponse::class.java)
        whenever(notOkResponse.isOk).thenReturn(false)
        assertThrows { instance.check(notOkResponse) }
                .isInstanceOf(FailedAssertionException::class.java)

    }

    companion object
    {

        fun runTheThing()
        {

        }
    }

}
